src/main/java/org/apache/lucene/replicator/nrt/FilteringSegmentInfosSearcherManager.java

```diff
@@ -0,0 +1,132 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.SegmentCommitInfo;
+import org.apache.lucene.index.SegmentInfos;
+import org.apache.lucene.index.SegmentReader;
+import org.apache.lucene.index.StandardDirectoryReader;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.ReferenceManager;
+import org.apache.lucene.search.SearcherFactory;
+import org.apache.lucene.search.SearcherManager;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.StringHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Extension of {@link SegmentInfosSearcherManager} which filters previous readers that are
+ * incompatible during a refresh. Useful for replicas, when a primary restart may lead to segment
+ * name reuse for un-committed data synced through nrt points.
+ */
+public class FilteringSegmentInfosSearcherManager extends SegmentInfosSearcherManager {
+  private static final Logger logger =
+      LoggerFactory.getLogger(FilteringSegmentInfosSearcherManager.class);
+  private final Directory dir;
+  private final Node node;
+  private final AtomicInteger openReaderCount = new AtomicInteger();
+  private final SearcherFactory searcherFactory;
+
+  public FilteringSegmentInfosSearcherManager(
+      Directory dir,
+      Node node,
+      ReferenceManager<IndexSearcher> mgr,
+      SearcherFactory searcherFactory)
+      throws IOException {
+    super(dir, node, ((SegmentInfosSearcherManager) mgr).getCurrentInfos(), searcherFactory);
+    this.dir = dir;
+    this.node = node;
+    if (searcherFactory == null) {
+      searcherFactory = new SearcherFactory();
+    }
+    this.searcherFactory = searcherFactory;
+  }
+
+  @Override
+  protected IndexSearcher refreshIfNeeded(IndexSearcher old) throws IOException {
+    final SegmentInfos newInfos = getCurrentInfos();
+    List<LeafReader> subs;
+    if (old == null) {
+      subs = null;
+    } else {
+      // create map of segment name to reader ordinal
+      final Map<String, Integer> oldReadersMap = new HashMap<>();
+      for (int i = 0; i < old.getIndexReader().leaves().size(); ++i) {
+        final SegmentReader sr = (SegmentReader) old.getIndexReader().leaves().get(i).reader();
+        oldReadersMap.put(sr.getSegmentName(), i);
+      }
+      subs = new ArrayList<>();
+      for (SegmentCommitInfo commitInfo : newInfos) {
+        Integer oldReaderIndex = oldReadersMap.get(commitInfo.info.name);
+        if (oldReaderIndex != null) {
+          SegmentReader oldReader =
+              (SegmentReader) old.getIndexReader().leaves().get(oldReaderIndex).reader();
```

Comment:
`indexReader.leaves()` doesn't look like a simple getter. Can we call it just once right after `oldReadersMap` is defined and store in a variable? There are checks that the IndexReader is still open within those function calls, but I doubt that the index will be closed during this.